File: js/segman_test.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { segment } = require('./segman.js');

function runTests() {
    // Read scenarios.jsonl from tests directory
    const scenariosPath = path.join(__dirname, '..', 'tests', 'scenarios.jsonl');
    const content = fs.readFileSync(scenariosPath, 'utf-8');
    const lines = content.split('\n').filter(line => line.trim() !== '');

    let scenarioCount = 0;
    let passCount = 0;
    const failures = [];

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        let scenario;

        try {
            scenario = JSON.parse(line);
        } catch (err) {
            console.error(`Line ${i + 1}: failed to parse JSON: ${err.message}`);
            continue;
        }

        scenarioCount++;

        // Run segmenter on context
        const sentences = segment(scenario.context);

        // Debug scenario 003
        if (scenario.id === '003') {
            console.log(`Context bytes: ${JSON.stringify(scenario.context)}`);
            console.log(`Expected bytes: ${JSON.stringify(scenario.expected)}`);
            sentences.forEach((s, idx) => {
                console.log(`Sentence[${idx}]: ${JSON.stringify(s)}`);
            });
        }

        // Check if expected sentence is in the output
        const found = sentences.some(sentence => sentence === scenario.expected);

        if (!found) {
            failures.push({
                id: scenario.id,
                context: scenario.context,
                expected: scenario.expected,
                got: sentences
            });
            console.log(`    segmenter_test.js:${i + 1}: Scenario ${scenario.id} FAILED`);
        } else {
            passCount++;
            console.log(`    segmenter_test.js:${i + 1}: Scenario ${scenario.id} PASSED`);
        }
    }

    console.log(`    segmenter_test.js:${lines.length + 1}: `);
    console.log(`        === Summary ===`);
    console.log(`        Total: ${scenarioCount}  Passed: ${passCount}  Failed: ${scenarioCount - passCount}`);

    // Print failures
    if (failures.length > 0) {
        console.log('\n=== FAILURES ===');
        failures.forEach(f => {
            console.log(`\nScenario ${f.id}:`);
            console.log(`  Context: ${JSON.stringify(f.context)}`);
            console.log(`  Expected: ${JSON.stringify(f.expected)}`);
            console.log(`  Got: ${JSON.stringify(f.got)}`);
        });
    }

    if (passCount === scenarioCount) {
        console.log('--- PASS: TestScenarios (0.00s)');
        console.log('PASS');
        console.log('ok  \tgithub.com/slackwing/segman/go\t0.002s');
        process.exit(0);
    } else {
        console.log('--- FAIL: TestScenarios (0.00s)');
        console.log('FAIL');
        process.exit(1);
    }
}

// Run tests if executed directly
if (require.main === module) {
    console.log('=== RUN   TestScenarios');
    runTests();
}

module.exports = { runTests };

File: src/segman/go/segman_test.go
package senseg

import (
	"bufio"
	"encoding/json"
	"os"
	"testing"
)

type Scenario struct {
	ID       string `json:"id"`
	Context  string `json:"context"`
	Expected string `json:"expected"`
}

func TestScenarios(t *testing.T) {
	// Read scenarios.jsonl from tests directory
	file, err := os.Open("../../../tests/scenarios.jsonl")
	if err != nil {
		t.Fatalf("Failed to open scenarios.jsonl: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lineNum := 0
	scenarioCount := 0
	passCount := 0

	for scanner.Scan() {
		lineNum++
		line := scanner.Text()

		// Skip empty lines
		if line == "" {
			continue
		}

		var scenario Scenario
		if err := json.Unmarshal([]byte(line), &scenario); err != nil {
			t.Errorf("Line %d: failed to parse JSON: %v", lineNum, err)
			continue
		}

		scenarioCount++

		// Run segmenter on context
		sentences := Segment(scenario.Context)

		// Debug scenario 003
		if scenario.ID == "003" {
			t.Logf("Context bytes: %q", scenario.Context)
			t.Logf("Expected bytes: %q", scenario.Expected)
			for i, s := range sentences {
				t.Logf("Sentence[%d]: %q", i, s)
			}
		}

		// Check if expected sentence is in the output
		found := false
		for _, sentence := range sentences {
			if sentence == scenario.Expected {
				found = true
				break
			}
		}

		if !found {
			t.Errorf("Scenario %s FAILED: expected sentence not found\n  Context: %q\n  Expected: %q\n  Got: %v",
				scenario.ID, scenario.Context, scenario.Expected, sentences)
		} else {
			passCount++
			t.Logf("Scenario %s PASSED", scenario.ID)
		}
	}

	if err := scanner.Err(); err != nil {
		t.Fatalf("Error reading scenarios.jsonl: %v", err)
	}

	t.Logf("\n=== Summary ===\nTotal: %d  Passed: %d  Failed: %d",
		scenarioCount, passCount, scenarioCount-passCount)

	if passCount < scenarioCount {
		t.Fail()
	}
}
